File: packages/desktop/main/service/TgrokService.ts
import { JsRequest } from "@drmer/core";
import { Tunnel } from "@tgrok/core";
import { JsService } from "./JsService";
import { tgrok } from "../tgrok";

class TgrokService extends JsService {

  public boot(req: JsRequest): void {
    tgrok.host = req.strParam("host");
    tgrok.port = req.numParam("port");
    tgrok.start([])
  }

  public reconnect(_req: JsRequest): void {
    tgrok.reconnect(true);
  }

  public open(req: JsRequest): void {
    tgrok.openTunnel(TgrokService.newTunnel(req.params));
  }

  private static newTunnel(conf: any) {
    return new Tunnel({
      id: conf.id,
      protocol: conf.protocol,
      hostname: "",
      subdomain: conf.subdomain,
      rport: 0,
      lhost: conf.localHost,
      lport: conf.localPort,
    });
  }

  public close(req: JsRequest): void {
    tgrok.closeTunnel(req.strParam("id"));
  }

  public remove(req: JsRequest): void {
    tgrok.removeTunnel(req.strParam("id"));
  }

  public status(_req: JsRequest): any {
    return tgrok.status;
  }

}

export {
  TgrokService,
};

File: packages/ios/Tgrok/service/TgrokService.swift
//
//  TgrokService.swift
//  Tgrok
//
//  Created by Yueyu Zhao on 2020/9/13.
//  Copyright © 2020 Yueyu Zhao. All rights reserved.
//

import Foundation
import SwiftyJSON

class TgrokService: JsService {
    
    static let shared = TgrokService()
    
    private override init() {
        super.init()
    }

    @objc func boot(_ req: JsRequest) {
        tgrok.boot(req.strParam("host"), UInt16(req.intParam("port")))
        DispatchQueue.main.async {
            tgrok.start([])
        }
    }

    @objc func reconnect(_ req: JsRequest) {
        req.callback(true)
    }
    
    @objc func open(_ req: JsRequest) {
        tgrok.openTunnel(Tunnel(req.params))
        req.callback(true)
    }
    
    @objc func close(_ req: JsRequest) {
        req.callback(tgrok.closeTunnel(req.strParam("id")))
    }
    
    @objc func remove(_ req: JsRequest) {
        req.callback(tgrok.removeTunnel(req.strParam("id")))
    }
    
    @objc func status(_ req: JsRequest) {
        req.callback(tgrok.status())
    }
    
}
